packages/block-editor/src/hooks/custom-fields.js

```diff
@@ -0,0 +1,136 @@
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+import { PanelBody, TextControl } from '@wordpress/components';
+import { __, sprintf } from '@wordpress/i18n';
+import { hasBlockSupport } from '@wordpress/blocks';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { useRef } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { InspectorControls } from '../components';
+import { useBlockEditingMode } from '../components/block-editing-mode';
+
+/**
+ * Filters registered block settings, extending attributes to include `connections`.
+ *
+ * @param {Object} settings Original block settings.
+ *
+ * @return {Object} Filtered block settings.
+ */
+export function addAttribute( settings ) {
+	if ( hasBlockSupport( settings, 'connections', true ) ) {
+		// Gracefully handle if settings.attributes is undefined.
+		settings.attributes = {
+			...settings.attributes,
+			connections: {
+				type: 'object',
+			},
+		};
+	}
+
+	return settings;
+}
+
+/**
+ * Override the default edit UI to include a new block inspector control for
+ * assigning a connection to blocks that has support for connections.
+ * Currently, only the `core/paragraph` block is supported and there is only a relation
+ * between paragraph content and a custom field.
+ *
+ * @param {WPComponent} BlockEdit Original component.
+ *
+ * @return {WPComponent} Wrapped component.
+ */
+export const withInspectorControl = createHigherOrderComponent(
+	( BlockEdit ) => {
+		return ( props ) => {
+			const blockEditingMode = useBlockEditingMode();
+			const hasCustomFieldsSupport = hasBlockSupport(
+				props.name,
+				'connections',
+				false
+			);
+			// We prevent that the content is lost when the user removes the custom field.
+			// Editing the content in the paragraph block with a placeholder is not the best solution.
+			const prevContent = useRef( props.attributes?.content );
+			if ( ! prevContent.current ) {
+				prevContent.current = '';
+			}
+			if ( hasCustomFieldsSupport && props.isSelected ) {
+				return (
+					<>
+						<BlockEdit { ...props } />
+						{ blockEditingMode === 'default' && (
+							<InspectorControls>
+								<PanelBody
+									title={ __( 'Connections' ) }
+									initialOpen={ true }
+								>
+									<TextControl
+										__nextHasNoMarginBottom
+										autoComplete="off"
+										label={ __( 'Custom field meta_key' ) }
```

Comment:
It would also be useful to mention that only 'string', 'boolean', 'integer', and 'number' fields are supported for now.

Reply:
What could the user add in a text input? I think it only allows strings.

Reply:
Sorry, I should have been more clear! I meant that you can only reference the meta fields that have a type `string`, `boolean`, `integer` or `number`. 

But perhaps it is too obvious that you cannot reference a field of type `object` or `array` here since this is a **Paragraph** block. What do you think?

Reply:
Ah sorry for not understanding it 😅  
In that case I guess is a good idea.

Reply:
Maybe not for this initial version, but I'd like to explore how to connect the paragraph content with a property of an object. If I am not mistaken, there are many meta fields that are object and you might want to connect one of its properties.

For example, if we want to connect the Image title attribute, it is stored in an `_wp_attachment_metadata` object with other properties like width or height.